packages/frontend-plugin-api/src/wiring/createExtension.ts

```diff
@@ -15,51 +15,68 @@
  */
 
 import { PortableSchema } from '../schema';
+import { ExtensionDataRef } from './createExtensionDataRef';
+import { ExtensionInput } from './createExtensionInput';
 import { BackstagePlugin } from './createPlugin';
-import { AnyExtensionDataMap, Extension } from './types';
 
 /** @public */
-export type ExtensionDataInputValues<
-  TInputs extends { [name in string]: { extensionData: AnyExtensionDataMap } },
-> = {
-  [InputName in keyof TInputs]: Array<
-    {
-      [DataName in keyof TInputs[InputName]['extensionData'] as TInputs[InputName]['extensionData'][DataName]['config'] extends {
-        optional: true;
-      }
-        ? never
-        : DataName]: TInputs[InputName]['extensionData'][DataName]['T'];
-    } & {
-      [DataName in keyof TInputs[InputName]['extensionData'] as TInputs[InputName]['extensionData'][DataName]['config'] extends {
-        optional: true;
-      }
-        ? DataName
-        : never]?: TInputs[InputName]['extensionData'][DataName]['T'];
-    }
-  >;
+export type AnyExtensionDataMap = {
+  [name in string]: ExtensionDataRef<unknown, { optional?: true }>;
 };
 
 /** @public */
-export type ExtensionDataBind<TMap extends AnyExtensionDataMap> = (
-  values: {
-    [DataName in keyof TMap as TMap[DataName]['config'] extends {
-      optional: true;
-    }
-      ? never
-      : DataName]: TMap[DataName]['T'];
-  } & {
-    [DataName in keyof TMap as TMap[DataName]['config'] extends {
-      optional: true;
-    }
-      ? DataName
-      : never]?: TMap[DataName]['T'];
-  },
-) => void;
+export type AnyExtensionInputMap = {
+  [inputName in string]: ExtensionInput<
+    AnyExtensionDataMap,
+    { optional: boolean; singleton: boolean }
+  >;
+};
+
+// TODO(Rugvip): This might be a quite useful utility type, maybe add to @backstage/types?
+/**
+ * Utility type to expand type aliases into their equivalent type.
+ * @ignore
+ */
+export type Expand<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;
```

Comment:
This helps converting this:

![image](https://github.com/backstage/backstage/assets/4984472/de211f29-4665-4feb-b395-16cac55f5dc2)

to this:

![image](https://github.com/backstage/backstage/assets/4984472/b2c53983-6baa-4abd-8b77-58ab4a1cee16)